shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,388 @@
+  return self->_drawableId;
+}
+
+- (CFTimeInterval)presentedTime {
+  return 0;
+}
+
+- (void)present {
+  [_layer presentTexture:self->_texture];
+  self->_presented = YES;
+}
+
+- (void)dealloc {
+  if (!_presented) {
+    [_layer returnTexture:self->_texture];
+  }
+}
+
+- (void)addPresentedHandler:(nonnull MTLDrawablePresentedHandler)block {
+}
+
+- (void)presentAtTime:(CFTimeInterval)presentationTime {
+}
+
+- (void)presentAfterMinimumDuration:(CFTimeInterval)duration {
+}
+
+@end
+
+@implementation FlutterMetalLayer
+
+@synthesize preferredDevice = _preferredDevice;
+@synthesize device = _device;
+@synthesize pixelFormat = _pixelFormat;
+@synthesize framebufferOnly = _framebufferOnly;
+@synthesize colorspace = _colorspace;
+@synthesize wantsExtendedDynamicRangeContent = _wantsExtendedDynamicRangeContent;
+
+- (instancetype)init {
+  if (self = [super init]) {
+    _preferredDevice = MTLCreateSystemDefaultDevice();
+    self.device = self.preferredDevice;
+    self.pixelFormat = MTLPixelFormatBGRA8Unorm;
+    _availableTextures = [[NSMutableSet alloc] init];
+
+    _displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(onDisplayLink:)];
+    [self setMaxRefreshRate:[DisplayLinkManager displayRefreshRate] forceMax:NO];
+    [_displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(didEnterBackground:)
+                                                 name:UIApplicationDidEnterBackgroundNotification
+                                               object:nil];
+  }
+  return self;
+}
+
+- (void)dealloc {
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)setMaxRefreshRate:(double)refreshRate forceMax:(BOOL)forceMax {
+  if (!DisplayLinkManager.maxRefreshRateEnabledOnIPhone) {
+    return;
+  }
+  double maxFrameRate = fmax(refreshRate, 60);
+  double minFrameRate = fmax(maxFrameRate / 2, 60);
+  if (@available(iOS 15.0, *)) {
+    _displayLink.preferredFrameRateRange =
+        CAFrameRateRangeMake(forceMax ? maxFrameRate : minFrameRate, maxFrameRate, maxFrameRate);
+  } else {
+    _displayLink.preferredFramesPerSecond = maxFrameRate;
+  }
+}
+
+- (void)onDisplayLink:(CADisplayLink*)link {
+  _didSetContentsDuringThisDisplayLinkPeriod = NO;
+  // Do not pause immediately, this seems to prevent 120hz while touching.
+  if (_displayLinkPauseCountdown == 3) {
+    _displayLink.paused = YES;
+    if (_displayLinkForcedMaxRate) {
+      [self setMaxRefreshRate:[DisplayLinkManager displayRefreshRate] forceMax:NO];
+      _displayLinkForcedMaxRate = NO;
+    }
+  } else {
+    ++_displayLinkPauseCountdown;
+  }
+}
+
+- (BOOL)isKindOfClass:(Class)aClass {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability-new"
+  // Pretend that we're a CAMetalLayer so that the rest of Flutter plays along
```

Comment:
Are there parts of flutter where this is needed?

Reply:
I copied the `"-Wunguarded-availability-new"` from some other part of engine. Everywhere CAMetalLayer is mentioned the simulator build will complain. Apparently it's an old well known bug.